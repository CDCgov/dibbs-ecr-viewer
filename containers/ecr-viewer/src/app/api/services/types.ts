import {
  ColumnType,
  Generated,
  Insertable,
  Selectable,
  Updateable,
} from "kysely";

export interface ecr_data {
  eICR_ID: Generated<string>;
  set_id: string | undefined;
  data_source: string | undefined;
  fhir_reference_link: string | undefined;
  eicr_version_number: string | undefined;

  patient_name_first: string;
  patient_name_last: string;
  patient_birth_date: ColumnType<Date, string | undefined, never>;

  date_created: ColumnType<Date, string | undefined, never>;
  report_date: ColumnType<Date, string>;
}

export interface ecr_rr_conditions {
  uuid: Generated<string>;
  eICR_ID: string;
  condition: string;
}

export interface ecr_rr_rule_summaries {
  uuid: Generated<string>;
  ecr_rr_conditions_id: string;
  rule_summary: string;
}

export type ECR = Selectable<ecr_data>;
export type NewECR = Insertable<ecr_data>;
export type ECRUpdate = Updateable<ecr_data>;

export type ECRConditions = Selectable<ecr_rr_conditions>;
export type NewECRConditions = Insertable<ecr_rr_conditions>;
export type ECRConditionsUpdate = Updateable<ecr_rr_conditions>;

export type ECRRuleSummaries = Selectable<ecr_rr_rule_summaries>;
export type NewECRRuleSummaries = Insertable<ecr_rr_rule_summaries>;
export type ECRRuleSummariesUpdate = Updateable<ecr_rr_rule_summaries>;

export interface Core {
  ecr_data: ecr_data;
  ecr_rr_conditions: ecr_rr_conditions;
  ecr_rr_rule_summaries: ecr_rr_rule_summaries;
}
