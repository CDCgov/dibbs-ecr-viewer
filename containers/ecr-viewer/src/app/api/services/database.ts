// Kysely ORM Connection Client

import { Kysely, PostgresDialect, MssqlDialect } from "kysely";
import { Pool } from "pg";
import * as tarn from "tarn";
import * as tedious from "tedious";

import { Extended } from "./extended_types";
import { Core } from "./types";

// Dialect to communicate with the database, interface to define its structure.

let db: Kysely<Core> | Kysely<Extended>;

if (process.env.METADATA_DATABASE_TYPE === "sqlserver") {
  db = new Kysely<Extended>({
    dialect: new MssqlDialect({
      tarn: {
        ...tarn,
        options: {
          min: 0,
          max: 100,
        },
      },
      tedious: {
        ...tedious,
        connectionFactory: () =>
          new tedious.Connection({
            authentication: {
              options: {
                password: process.env.SQL_SERVER_PASSWORD,
                userName: process.env.SQL_SERVER_USER || "sa",
              },
              type: "default",
            },
            options: {
              database: "master",
              port: 1433,
              trustServerCertificate: true,
              connectTimeout: 30000,
            },
            server: process.env.SQL_SERVER_HOST || "localhost",
          }),
      },
    }),
  });
} else {
  db = new Kysely<Core>({
    dialect: new PostgresDialect({
      pool: new Pool({
        database: process.env.POSTGRES_DATABASE || "ecr_viewer_db",
        host: process.env.POSTGRES_HOST,
        user: process.env.POSTGRES_USER,
        port: parseInt(process.env.POSTGRES_PORT || "5432"),
        max: parseInt(process.env.POSTGRES_MAX_THREADPOOL || "10"),
      }),
    }),
  });
}

export { db };
