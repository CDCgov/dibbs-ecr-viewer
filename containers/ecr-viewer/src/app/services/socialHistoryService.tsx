import { Bundle, Observation } from "fhir/r4";
import { PathMappings, noData } from "../utils/data-utils";

import { evaluate } from "../utils/evaluate";
import { evaluateValue } from "./evaluateFhirDataService";
import { JsonTable } from "../view-data/components/common";
import { formatDate } from "./formatDateService";
import { TableRow } from "./htmlTableService";

/**
 * Extracts travel history information from the provided FHIR bundle based on the FHIR path mappings.
 * @param fhirBundle - The FHIR bundle containing patient travel history data.
 * @param mappings - An object containing the FHIR path mappings.
 * @returns - A formatted table representing the patient's travel history, or undefined if no relevant data is found.
 */
export const evaluateTravelHistoryTable = (
  fhirBundle: Bundle,
  mappings: PathMappings,
) => {
  const travelHistory: Observation[] = evaluate(
    fhirBundle,
    mappings.patientTravelHistory,
  );

  const columnInfo = [
    {
      columnName: "Location",
      infoPath: "travelHistoryLocation",
    },
    {
      columnName: "Start Date",
      infoPath: "travelHistoryStartDate",
      applyToValue: formatDate,
    },
    {
      columnName: "End Date",
      infoPath: "travelHistoryEndDate",
      applyToValue: formatDate,
    },
    {
      columnName: "Purpose",
      infoPath: "travelHistoryPurpose",
    },
  ];

  const tables = travelHistory
    .map((act) => {
      return columnInfo.reduce(
        (row, { columnName, infoPath, applyToValue }) => {
          let val = evaluateValue(act, mappings[infoPath]);
          if (applyToValue) {
            val = applyToValue(val) ?? "";
          }
          return { ...row, [columnName]: { value: val || noData } };
        },
        {} as TableRow[],
      );
    })
    .filter((row) =>
      Object.values(row).some((v) => (v.value as any) !== noData),
    );

  if (!tables.flat().length) return undefined;

  return (
    <JsonTable
      rawTable={{ tables, resultName: "Travel History" }}
      outerBorder
      className="caption-data-title margin-y-0"
    />
  );
};
