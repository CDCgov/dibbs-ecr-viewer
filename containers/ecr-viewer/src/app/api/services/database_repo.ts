import { Kysely } from "kysely";

import { db } from "./database";
import {
  ECRConditions,
  NewECRConditions,
  ECRConditionsUpdate,
  ECRRuleSummaries,
  NewECRRuleSummaries,
  ECRRuleSummariesUpdate,
  ECR,
  NewECR,
  ECRUpdate,
} from "./types";
import { Core } from "./types";

// ECR_DATA

/**
 * Finds an eICR by its ID
 * @async
 * @function findEcrById
 * @param id - thd ID of the eCR to be looked up
 * @returns an eICR object
 */
export async function findEcrById(id: string | null) {
  if (!id) {
    return console.error("eICR ID is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .selectFrom("ecr_data")
      .where("eICR_ID", "=", id)
      .selectAll()
      .executeTakeFirst();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Finds an eICR by its criteria
 * @async
 * @function findEcr
 * @param criteria - the Partial<ECR> used to filter against
 * @returns an eICR object
 */
export async function findEcr(criteria: Partial<ECR> | null) {
  let query = (db as Kysely<Core>).selectFrom("ecr_data");

  if (!criteria || criteria === null) {
    return console.error("eICR Criteria is required.");
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.set_id) {
    query = query.where("set_id", "=", criteria.set_id);
  }

  if (criteria.data_source) {
    query = query.where("data_source", "=", criteria.data_source);
  }

  if (criteria.fhir_reference_link) {
    query = query.where(
      "fhir_reference_link",
      "=",
      criteria.fhir_reference_link,
    );
  }

  if (criteria.patient_name_last !== undefined) {
    query = query.where(
      "patient_name_last",
      criteria.patient_name_last === null ? "is" : "=",
      criteria.patient_name_last,
    );
  }

  if (criteria.patient_birth_date) {
    query = query.where("patient_birth_date", "=", criteria.patient_birth_date);
  }

  if (criteria.date_created) {
    query = query.where("date_created", "=", criteria.date_created);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eICR object
 * @async
 * @function createEcr
 * @param ecr - the NewECR to be persisted
 * @returns the created eICR object
 */
export async function createEcr(ecr: NewECR | null) {
  if (!ecr || ecr === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_data")
      .values(ecr)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eICR object
 * @async
 * @function updateEcr
 * @param eICR_ID - the eICR_ID of the eCR to be updated
 * @param updateWith - the ECRUpdate to be applied to the existing record
 * @returns the updated eICR object
 */
export async function updateEcr(eICR_ID: string | null, updateWith: ECRUpdate) {
  await (db as Kysely<Core>)
    .updateTable("ecr_data")
    .set(updateWith)
    .where("eICR_ID", "=", eICR_ID)
    .execute();
}

/**
 * Deletes an eICR object
 * @async
 * @function deleteEcr
 * @param eICR_ID - the eICR_ID of the record to be deleted
 * @returns the deleted eICR object
 */
export async function deleteEcr(eICR_ID: string | null) {
  const ecr = await findEcrById(eICR_ID);

  if (ecr) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_data")
      .where("eICR_ID", "=", eICR_ID)
      .execute();
  }

  return ecr;
}

// ECR_RR_CONDITIONS

/**
 * Finds an eCR condition by its ID
 * @async
 * @function findEcrConditionById
 * @param id - the UUID of the eCR condition being looked up
 * @returns an eCR condition object
 */
export async function findEcrConditionById(id: string) {
  return await (db as Kysely<Core>)
    .selectFrom("ecr_rr_conditions")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR condition by its criteria
 * @async
 * @function findEcrCondition
 * @param criteria - the Partial<ECRConditions> filtering criteria to be applied
 * @returns an eCR condition object
 */
export async function findEcrCondition(criteria: Partial<ECRConditions>) {
  let query = (db as Kysely<Core>).selectFrom("ecr_rr_conditions");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.condition) {
    query = query.where("condition", "=", criteria.condition);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR condition object
 * @async
 * @function createEcrCondition
 * @param condition - the NewECRConditions to be created
 * @returns the created eCR condition object
 */
export async function createEcrCondition(condition: NewECRConditions | null) {
  if (!condition || condition === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_rr_conditions")
      .values(condition)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR condition object
 * @async
 * @function updateEcrCondition
 * @param uuid - the UUID of the eCR Condition to be updated
 * @param updateWith - the ECRConditionsUpdate to be applied
 * @returns the updated eCR condition object
 */
export async function updateEcrCondition(
  uuid: string,
  updateWith: ECRConditionsUpdate,
) {
  await (db as Kysely<Core>)
    .updateTable("ecr_rr_conditions")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR condition object
 * @async
 * @function deleteEcrCondition
 * @param uuid - thge UUID of the record to be deleted
 * @returns the deleted eCR condition object
 */
export async function deleteEcrCondition(uuid: string) {
  const ecr = await findEcrConditionById(uuid);

  if (ecr) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_rr_conditions")
      .where("uuid", "=", uuid)
      .execute();
  }

  return ecr;
}

// ECR_RR_RULE_SUMMARIES

/**
 * Finds an eCR rule summary by its ID
 * @async
 * @function findEcrRuleById
 * @param id - the ID of the eCR being looked up
 * @returns an eCR rule object
 */
export async function findEcrRuleById(id: string) {
  return await (db as Kysely<Core>)
    .selectFrom("ecr_rr_rule_summaries")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR rule summary by its criteria
 * @async
 * @function findEcrRule
 * @param criteria - the Partial<ECRRuleSummaries> filtering criteria
 * @returns an eCR rule object
 */
export async function findEcrRule(criteria: Partial<ECRRuleSummaries>) {
  let query = (db as Kysely<Core>).selectFrom("ecr_rr_rule_summaries");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.ecr_rr_conditions_id) {
    query = query.where(
      "ecr_rr_conditions_id",
      "=",
      criteria.ecr_rr_conditions_id,
    );
  }

  if (criteria.rule_summary) {
    query = query.where("rule_summary", "=", criteria.rule_summary);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR rule summary object
 * @async
 * @param rule_summary - the NewECRRuleSummaries record to be created
 * @function createEcrRule
 * @returns the created eCR rule object
 */
export async function createEcrRule(rule_summary: NewECRRuleSummaries) {
  if (!rule_summary || rule_summary === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Core>)
      .insertInto("ecr_rr_rule_summaries")
      .values(rule_summary)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR rule summary object
 * @async
 * @function updateEcrRule
 * @param uuid - the UUID of the eCR rule to be updated
 * @param updateWith - the ECRRuleSummariesUpdate to be applied to the record
 * @returns the updated eCR rule object
 */
export async function updateEcrRule(
  uuid: string,
  updateWith: ECRRuleSummariesUpdate,
) {
  await (db as Kysely<Core>)
    .updateTable("ecr_rr_rule_summaries")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR rule summary object
 * @async
 * @function deleteEcrRule
 * @param uuid - the UUID of the eCR rule to be deleted
 * @returns the deleted eCR rule object
 */
export async function deleteEcrRule(uuid: string) {
  const rule_summary = await findEcrRuleById(uuid);

  if (rule_summary) {
    await (db as Kysely<Core>)
      .deleteFrom("ecr_rr_rule_summaries")
      .where("uuid", "=", uuid)
      .execute();
  }

  return rule_summary;
}
