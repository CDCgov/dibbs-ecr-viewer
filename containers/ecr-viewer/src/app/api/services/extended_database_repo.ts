import { Kysely } from "kysely";

import { db } from "./database";
import {
  Extended,
  ExtendedECR,
  NewExtendedECR,
  ExtendedECRUpdate,
  PatientAddress,
  NewPatientAddress,
  PatientAddressUpdate,
  ECRLabs,
  NewECRLabs,
  ECRLabsUpdate,
  ECRConditions,
  NewECRConditions,
  ECRConditionsUpdate,
  ECRRuleSummaries,
  NewECRRuleSummaries,
  ECRRuleSummariesUpdate,
} from "./extended_types";

/**
 * Finds an eICR by its ID
 * @async
 * @function findEcrById
 * @param id - the ID of the ExtendedEcr being looked up
 * @returns an eICR object
 */
export async function findExtendedEcrById(id: string | null) {
  if (!id) {
    return console.error("eICR ID is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .selectFrom("ecr_data")
      .where("eICR_ID", "=", id)
      .selectAll()
      .executeTakeFirst();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Finds an eICR by its given criteria
 * @async
 * @function findExtendedEcr
 * @param criteria - the Partial<ExtendedECR> filtering criteria
 * @returns an eICR object
 */
export async function findExtendedEcr(criteria: Partial<ExtendedECR> | null) {
  let query = (db as Kysely<Extended>).selectFrom("ecr_data");

  if (!criteria || criteria === null) {
    return console.error("eICR Criteria is required.");
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.set_id) {
    query = query.where("set_id", "=", criteria.set_id);
  }

  if (criteria.fhir_reference_link) {
    query = query.where("set_id", "=", criteria.fhir_reference_link);
  }

  if (criteria.last_name !== undefined) {
    query = query.where(
      "last_name",
      criteria.last_name === null ? "is" : "=",
      criteria.last_name,
    );
  }

  if (criteria.first_name !== undefined) {
    query = query.where("first_name", "=", criteria.first_name);
  }

  if (criteria.fhir_reference_link) {
    query = query.where(
      "fhir_reference_link",
      "=",
      criteria.fhir_reference_link,
    );
  }

  if (criteria.last_name !== undefined) {
    query = query.where(
      "last_name",
      criteria.last_name === null ? "is" : "=",
      criteria.last_name,
    );
  }

  if (criteria.birth_date) {
    query = query.where("birth_date", "=", criteria.birth_date);
  }

  if (criteria.date_created) {
    query = query.where("date_created", "=", criteria.date_created);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eICR object
 * @async
 * @function createEcr
 * @param ecr - the NewExtendedECR to be persisted
 * @returns the created eICR object
 */
export async function createExtendedEcr(ecr: NewExtendedECR | null) {
  if (!ecr || ecr === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .insertInto("ecr_data")
      .values(ecr)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eICR object
 * @async
 * @function updateEcr
 * @param eICR_ID - the ID of the eICR to be updated
 * @param updateWith - the ECRUpdate to be applied to the existing record
 * @returns the updated eICR object
 */
export async function updateExtendedEcr(
  eICR_ID: string | null,
  updateWith: ExtendedECRUpdate,
) {
  await (db as Kysely<Extended>)
    .updateTable("ecr_data")
    .set(updateWith)
    .where("eICR_ID", "=", eICR_ID)
    .execute();
}

/**
 * Deletes an eICR object
 * @async
 * @function deleteEcr
 * @param eICR_ID - the ID of the eICR to be deleted
 * @returns the deleted eICR object
 */
export async function deleteExtendedEcr(eICR_ID: string | null) {
  const ecr = await findExtendedEcrById(eICR_ID);

  if (ecr) {
    await (db as Kysely<Extended>)
      .deleteFrom("ecr_data")
      .where("eICR_ID", "=", eICR_ID)
      .execute();
  }

  return ecr;
}

// PATIENT_ADDRESS

/**
 * Finds a patient_address by its ID
 * @async
 * @function findAddressById
 * @param id - the ID of the address record
 * @returns a patient_address object
 */
export async function findAddressById(id: string) {
  return await (db as Kysely<Extended>)
    .selectFrom("patient_address")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds a patient_address by its criteria
 * @async
 * @function findAddress
 * @param criteria - the Partial<PatientAddress> filtering criteria to be looked up
 * @returns a patient_address object
 */
export async function findAddress(criteria: Partial<PatientAddress>) {
  let query = (db as Kysely<Extended>).selectFrom("patient_address");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  return await query.selectAll().execute();
}

/**
 * Creates a patient_address object
 * @async
 * @function createAddress
 * @param patient_address - the NewPatientAddress to be persisted
 * @returns the created patient_address object
 */
export async function createAddress(patient_address: NewPatientAddress) {
  if (!patient_address || patient_address === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .insertInto("patient_address")
      .values(patient_address)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates a patient_address object
 * @async
 * @function updateAddress
 * @param uuid - the UUID of of the record to be updated
 * @param updateWith - the PatientAddressUpdate to be applied to the existing record
 * @returns the updated patient_address object
 */
export async function updateAddress(
  uuid: string,
  updateWith: PatientAddressUpdate,
) {
  await (db as Kysely<Extended>)
    .updateTable("patient_address")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes a patient_address object
 * @async
 * @function deleteAddress
 * @param uuid - the UUID of the record to be deleted
 * @returns the deleted patient_address object
 */
export async function deleteAddress(uuid: string) {
  const address = await findAddressById(uuid);

  if (address) {
    await (db as Kysely<Extended>)
      .deleteFrom("patient_address")
      .where("uuid", "=", uuid)
      .execute();
  }

  return address;
}

// ECR_LABS

/**
 * Finds an eCR Lab by its ID
 * @async
 * @function findLabById
 * @param id - the ID of the lab being looked up
 * @returns an eCR Lab object
 */
export async function findLabById(id: string) {
  return await (db as Kysely<Extended>)
    .selectFrom("ecr_labs")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR Lab by its criteria
 * @async
 * @function findLab
 * @param criteria - the Partial<ECRLabs> filtering criteria
 * @returns an eCR Lab object
 */
export async function findLab(criteria: Partial<ECRLabs>) {
  let query = (db as Kysely<Extended>).selectFrom("ecr_labs");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.test_type) {
    query = query.where("test_type", "=", criteria.test_type);
  }

  if (criteria.test_type) {
    query = query.where("test_type", "=", criteria.test_type);
  }

  if (criteria.test_type_code) {
    query = query.where("test_type_code", "=", criteria.test_type_code);
  }

  if (criteria.test_type_system) {
    query = query.where("test_type_system", "=", criteria.test_type_system);
  }

  if (criteria.test_result_qualitative) {
    query = query.where(
      "test_result_qualitative",
      "=",
      criteria.test_result_qualitative,
    );
  }

  if (criteria.test_result_quantitative !== undefined) {
    query = query.where(
      "test_result_quantitative",
      "=",
      criteria.test_result_quantitative,
    );
  }

  if (criteria.test_result_units) {
    query = query.where("test_result_units", "=", criteria.test_result_units);
  }

  if (criteria.test_result_code) {
    query = query.where("test_result_code", "=", criteria.test_result_code);
  }

  if (criteria.test_result_code_display) {
    query = query.where(
      "test_result_code_display",
      "=",
      criteria.test_result_code_display,
    );
  }

  if (criteria.test_result_code_system) {
    query = query.where(
      "test_result_code_system",
      "=",
      criteria.test_result_code_system,
    );
  }

  if (criteria.test_result_interpretation) {
    query = query.where(
      "test_result_interpretation",
      "=",
      criteria.test_result_interpretation,
    );
  }

  if (criteria.test_result_interpretation_code) {
    query = query.where(
      "test_result_interpretation_code",
      "=",
      criteria.test_result_interpretation_code,
    );
  }

  if (criteria.test_result_interpretation_system) {
    query = query.where(
      "test_result_interpretation_system",
      "=",
      criteria.test_result_interpretation_system,
    );
  }

  if (criteria.test_result_reference_range_low_value !== undefined) {
    // Check for null as well
    query = query.where(
      "test_result_reference_range_low_value",
      "=",
      criteria.test_result_reference_range_low_value,
    );
  }

  if (criteria.test_result_reference_range_low_units) {
    query = query.where(
      "test_result_reference_range_low_units",
      "=",
      criteria.test_result_reference_range_low_units,
    );
  }

  if (criteria.test_result_reference_range_high_value !== undefined) {
    // Check for null as well
    query = query.where(
      "test_result_reference_range_high_value",
      "=",
      criteria.test_result_reference_range_high_value,
    );
  }

  if (criteria.test_result_reference_range_high_units) {
    query = query.where(
      "test_result_reference_range_high_units",
      "=",
      criteria.test_result_reference_range_high_units,
    );
  }

  if (criteria.specimen_type) {
    query = query.where("specimen_type", "=", criteria.specimen_type);
  }

  if (criteria.specimen_collection_date) {
    query = query.where(
      "specimen_collection_date",
      "=",
      criteria.specimen_collection_date,
    );
  }

  if (criteria.performing_lab) {
    query = query.where("performing_lab", "=", criteria.performing_lab);
  }
  return await query.selectAll().execute();
}

/**
 * Creates an eCR Lab object
 * @async
 * @param lab - the NewECRLabs to be persisted
 * @function createLab
 * @returns the created eCR Lab object
 */
export async function createLab(lab: NewECRLabs | null) {
  if (!lab || lab === null) {
    return console.error("eICR Lab Data is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .insertInto("ecr_labs")
      .values(lab)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR Lab object
 * @async
 * @function updateLab
 * @param uuid - the UUID of the record to be updated
 * @param updateWith - the ECRLapsUpdate to be applied to the existing record
 * @returns the updated eCR Lab object
 */
export async function updateLab(uuid: string, updateWith: ECRLabsUpdate) {
  await (db as Kysely<Extended>)
    .updateTable("ecr_labs")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR condition object
 * @async
 * @function deleteLab
 * @param uuid - the UUID of the record to be deleted
 * @returns the deleted eCR Lab object
 */
export async function deleteLab(uuid: string) {
  const ecr = await findLabById(uuid);

  if (ecr) {
    await (db as Kysely<Extended>)
      .deleteFrom("ecr_labs")
      .where("uuid", "=", uuid)
      .execute();
  }

  return ecr;
}

// ECR_RR_CONDITIONS

/**
 * Finds an eCR condition by its ID
 * @async
 * @function findEcrConditionById
 * @param id - the ID of the record being looked up
 * @returns an eCR condition object
 */
export async function findEcrConditionById(id: string) {
  return await (db as Kysely<Extended>)
    .selectFrom("ecr_rr_conditions")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR condition by its criteria
 * @async
 * @function findEcrCondition
 * @param criteria - the Partial<ECRConditions> filter to be looked up
 * @returns an eCR condition object
 */
export async function findEcrCondition(criteria: Partial<ECRConditions>) {
  let query = (db as Kysely<Extended>).selectFrom("ecr_rr_conditions");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.eICR_ID) {
    query = query.where("eICR_ID", "=", criteria.eICR_ID);
  }

  if (criteria.condition) {
    query = query.where("condition", "=", criteria.condition);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR condition object
 * @async
 * @function createEcrCondition
 * @param condition - the NewECRConditions to be persisted
 * @returns the created eCR condition object
 */
export async function createEcrCondition(condition: NewECRConditions | null) {
  if (!condition || condition === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .insertInto("ecr_rr_conditions")
      .values(condition)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR condition object
 * @async
 * @function updateEcrCondition
 * @param uuid - the UUID of the record to be updated
 * @param updateWith - the ECRConditionsUpdate to be applied to the existing record
 * @returns the updated eCR condition object
 */
export async function updateEcrCondition(
  uuid: string,
  updateWith: ECRConditionsUpdate,
) {
  await (db as Kysely<Extended>)
    .updateTable("ecr_rr_conditions")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR condition object
 * @async
 * @function deleteEcrCondition
 * @param uuid - the UUID of the record to be deleted
 * @returns the deleted eCR condition object
 */
export async function deleteEcrCondition(uuid: string) {
  const ecr = await findEcrConditionById(uuid);

  if (ecr) {
    await (db as Kysely<Extended>)
      .deleteFrom("ecr_rr_conditions")
      .where("uuid", "=", uuid)
      .execute();
  }

  return ecr;
}

// ECR_RR_RULE_SUMMARIES

/**
 * Finds an eCR rule summary by its ID
 * @async
 * @function findEcrRuleById
 * @param id - the ID of the record being looked up
 * @returns an eCR rule object
 */
export async function findEcrRuleById(id: string) {
  return await (db as Kysely<Extended>)
    .selectFrom("ecr_rr_rule_summaries")
    .where("uuid", "=", id)
    .selectAll()
    .executeTakeFirst();
}

/**
 * Finds an eCR rule summary by its criteria
 * @async
 * @function findEcrRule
 * @param criteria - the Parcial<ECRRuleSummaries> to filter the record being looked up
 * @returns an eCR rule object
 */
export async function findEcrRule(criteria: Partial<ECRRuleSummaries>) {
  let query = (db as Kysely<Extended>).selectFrom("ecr_rr_rule_summaries");

  if (criteria.uuid) {
    query = query.where("uuid", "=", criteria.uuid);
  }

  if (criteria.ecr_rr_conditions_id) {
    query = query.where(
      "ecr_rr_conditions_id",
      "=",
      criteria.ecr_rr_conditions_id,
    );
  }

  if (criteria.rule_summary) {
    query = query.where("rule_summary", "=", criteria.rule_summary);
  }

  return await query.selectAll().execute();
}

/**
 * Creates an eCR rule summary object
 * @async
 * @function createEcrRule
 * @param rule_summary - the NewECRRuleSummaries to be persisted
 * @returns the created eCR rule object
 */
export async function createEcrRule(rule_summary: NewECRRuleSummaries) {
  if (!rule_summary || rule_summary === null) {
    return console.error("eICR Data is required.");
  }
  try {
    return await (db as Kysely<Extended>)
      .insertInto("ecr_rr_rule_summaries")
      .values(rule_summary)
      .returningAll()
      .executeTakeFirstOrThrow();
  } catch (error) {
    console.error(error);
  }
}

/**
 * Updates an eCR rule summary object
 * @async
 * @function updateEcrRule
 * @param uuid - the UUID of the record being updated
 * @param updateWith - the ECRRuleSummariesUpdate to be applied to the existing record
 * @returns the updated eCR rule object
 */
export async function updateEcrRule(
  uuid: string,
  updateWith: ECRRuleSummariesUpdate,
) {
  await (db as Kysely<Extended>)
    .updateTable("ecr_rr_rule_summaries")
    .set(updateWith)
    .where("uuid", "=", uuid)
    .execute();
}

/**
 * Deletes an eCR rule summary object
 * @async
 * @function deleteEcrRule
 * @param uuid - the UUID of the record to be deleted
 * @returns the deleted eCR rule object
 */
export async function deleteEcrRule(uuid: string) {
  const rule_summary = await findEcrRuleById(uuid);

  if (rule_summary) {
    await (db as Kysely<Extended>)
      .deleteFrom("ecr_rr_rule_summaries")
      .where("uuid", "=", uuid)
      .execute();
  }

  return rule_summary;
}
