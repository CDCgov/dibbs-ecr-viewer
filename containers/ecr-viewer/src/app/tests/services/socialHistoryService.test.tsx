import { evaluateTravelHistoryTable } from "@/app/services/socialHistoryService";
import BundleWithTravelHistory from "../assets/BundleTravelHistory.json";
import { Bundle } from "fhir/r4";
import { render } from "@testing-library/react";

describe("Travel History", () => {
  it("should display a table ", () => {
    const { container } = render(
      evaluateTravelHistoryTable(BundleWithTravelHistory as unknown as Bundle),
    );
    expect(container).toMatchSnapshot();
  });
  it("should display nothing when no travel history is available", () => {
    expect(evaluateTravelHistoryTable({} as Bundle)).toBeUndefined();
  });
});
