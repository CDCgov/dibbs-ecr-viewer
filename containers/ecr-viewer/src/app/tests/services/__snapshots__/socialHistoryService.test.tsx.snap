// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Travel History should display a table  1`] = `
<div>
  <table
    class="usa-table usa-table--borderless width-full table-caption-margin caption-normal-weight margin-bottom-2 caption-data-title margin-y-0 border-top border-left border-right"
    data-testid="table"
  >
    <caption>
      Travel History
    </caption>
    <thead>
      <tr>
        <th
          class="tableHeader bg-gray-5 minw-10"
          scope="col"
        >
          Location
        </th>
        <th
          class="tableHeader bg-gray-5 minw-10"
          scope="col"
        >
          Start Date
        </th>
        <th
          class="tableHeader bg-gray-5 minw-10"
          scope="col"
        >
          End Date
        </th>
        <th
          class="tableHeader bg-gray-5 minw-10"
          scope="col"
        >
          Purpose
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Traveled to Naboo, Tatooine and Alderaan
        </td>
        <td>
          01/18/1888
        </td>
        <td>
          02/18/1898
        </td>
        <td>
          Active duty military (occupation)
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
